network/openssl_privatekey.py

```diff
@@ -0,0 +1,225 @@
+        required: true
+        description:
+            - Name of the folder in which the generated SSL private key will be written
+'''
+
+EXAMPLES = '''
+# Generate an OpenSSL private key with the default values (4096 bits, RSA)
+- openssl_privatekey: name=ansible.com.pem path=/etc/ssl/private
+
+# Generate an OpenSSL private key with a different size (2048 bits)
+- openssl_privatekey: name=ansible.com.pem path=/etc/ssl/private size=2048
+
+# Force regenerate an OpenSSL private key if it already exists
+- openssl_privatekey: name=ansible.com.pem path=/etc/ssl/private force=True
+
+# Generate an OpenSSL private key with a different algorithm (DSA)
+- openssl_privatekey: name=ansible.com.pem path=/etc/ssl/private type=DSA
+'''
+
+RETURN = '''
+size:
+    description: Size of the SSL private key
+    returned:
+        - changed
+        - success
+    type: integer
+    sample: 4096
+type:
+    description: Algorithm used to generate the SSL private key
+    returned:
+        - changed
+        - success
+    type: string
+    sample: RSA
+privatekey:
+    description: Path to the generated SSL private key
+    returned:
+        - changed
+        - success
+    type: string
+    sample: /etc/ssl/private/ansible.com.pem
+'''
+
+class PrivateKeyError(Exception):
+    pass
+
+
+class PrivateKey(object):
+
+    def __init__(self, module):
+        self.size = module.params['size']
+        self.state = module.params['state']
+        self.name = module.params['name']
+        self.type = module.params['type']
+        self.force = module.params['force']
+        self.path = module.params['path']
+        self.changed = True
+
+    @property
+    def file_path(self):
+        return os.path.join(self.path, self.name)
+   
+    def generate(self):
+        '''Generate a keypair.'''
+
+        if not self.force and os.path.exists(self.file_path):
+            self.changed = False
+            return
+
+        self.privatekey = crypto.PKey()
+
+        if self.type == 'RSA':
+            crypto_type = crypto.TYPE_RSA
+        else:
+            crypto_type = crypto.TYPE_DSA
+
+        try:
+            self.privatekey.generate_key(crypto_type, self.size)
+            privatekey_file = open(self.file_path, 'w')
+            privatekey_file.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
+            privatekey_file.close()
+        except TypeError:
+            raise PrivateKeyError()
+        except ValueError:
+            raise PrivateKeyError()
+        except IOError:
+            try:
+                os.remove(self.file_path)
+            except OSError:
+                pass
+            raise PrivateKeyError()
+
+    def remove(self):
+        '''Remove the private key from the filesystem.'''
+        if os.path.exists(self.file_path):
+            try:
+                os.remove(self.file_path)
+            except OSError:
+                raise PrivateKeyError()
+        else:
+            self.changed = False
```

Comment:
done.

Thank you for the pointer about `get_exception()`. It is good to know it exists.